### PR TITLE
DAOS-11227 pool: Retry map_refresh on more errors

### PR DESCRIPTION
**Pull request by liw**
POOL_TGT_QUERY_MAP RPCs may encounter a large number of different remote
errors. For instance, from a restarted-but-not-yet-reintegrated engine,
we may get -DER_NO_HDL; Zhao Zhen has also observed -DER_OOG. For those
errors that are not normally retryable, this patch lets map_refresh
retry a limited number of times and fall back to dc_pool_query:

  - To use dc_pool_query, dc_pool_create_map_refresh_task has to take a
    pool handle instead of a dc_pool object.

  - Tune the backoff sequence of a map_refresh task a bit for hopefully
    better scalability.

  - Add a new daos_test case for the new fallback mechanism. And, bump
    the corresponding test timeout a little, since this new test
    involves a rebuild/reintegration cycle.

Signed-off-by: Li Wei <wei.g.li@intel.com>
Required-githooks: true

### CODE REVIEW

File: src/pool/cli.c [hunk 1]
```diff
@@ -1763,12 +1798,16 @@ map_refresh_cb(tse_task_t *task, void *varg)
...
+			DP_RC(rc));
 		reinit = true;
 		goto out;
 	} else if (rc != 0) {
-		D_ERROR(DF_UUID": failed to fetch pool map: "DF_RC"\n",
-			DP_UUID(pool->dp_pool), DP_RC(rc));
+		D_DEBUG(DB_MD, DF_UUID": %p: serious: "DF_RC"\n", DP_UUID(pool->dp_pool), task,
+			DP_RC(rc));
+		arg->mra_n--;
+		reinit = true;
```
Comment (liuxuezhao) [inline src/pool/cli.c:1810]: will it finally cause fall through (create new independent pool query task) again and again for some un-retryable error? for example DER_NOMEM or other un-retryable err?

**Review by liuxuezhao — APPROVED** [review]

File: src/pool/cli.c (hunk 1 above)
Comment (liw) [reply src/pool/cli.c:1810]: If rc == -DER_NOMEM here, we know that (1) it is from the reply buffer and (2) the engine side has encountered a memory allocation failure, and we decrement mra_n and rerun the map_refresh to choose another engine. When choosing another engine, if mra_n has reached 0, we fall back to dc_pool_query, in which case we don't use or run map_refresh_cb anymore.

Comment (liuxuezhao) [reply src/pool/cli.c:1810]: ok, thanks

**Review by daltonbohning — COMMENTED** [review]
ftest LGTM

**Review by wangdi1 — APPROVED** [review]

File: src/pool/cli.c [hunk 2]
```diff
@@ -1929,11 +1976,50 @@ map_refresh(tse_task_t *task)
...
-		pool->dp_map_task = task;
+		register_map_task(pool, task);
 	}
 
 	rank = choose_map_refresh_rank(arg);
+	if (rank == CRT_NO_RANK) {
+		tse_task_t	       *query_task;
+		daos_pool_query_t      *query_arg;
+
+		/* Fall back to dc_pool_query. */
```
[the remark this answers is not in the document]
Comment (liw) [reply src/pool/cli.c:1987]: Besides the destination of the request, POOL_QUERY is a PS request, and it needs to go through an rsvc_client, that is, rsvc_client_choose, rsvc_client_complete_rpc, etc. I was worried that making the map_refresh code work in both modes might be confusing.

**Pull request merged**